openedx/features/lti_course_tab/tab.py

```diff
@@ -0,0 +1,245 @@
+from typing import Dict
+from urllib.parse import quote
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.contrib.sites.shortcuts import get_current_site
+from django.http import HttpRequest
+from django.utils.translation import get_language, to_locale, ugettext_lazy
+from lti_consumer.lti_1p1.contrib.django import lti_embed
+from lti_consumer.models import LtiConfiguration
+from opaque_keys.edx.keys import CourseKey
+from web_fragments.fragment import Fragment
+
+from lms.djangoapps.courseware.access import get_user_role
+from lms.djangoapps.courseware.tabs import EnrolledTab
+from openedx.core.djangoapps.discussions.models import DiscussionsConfiguration
+from openedx.core.djangolib.markup import HTML
+from common.djangoapps.student.models import anonymous_id_for_user
+from xmodule.course_module import CourseDescriptor
+from xmodule.tabs import TabFragmentViewMixin, key_checker
+
+
+class LtiCourseLaunchMixin(TabFragmentViewMixin):
+    """
+    Mixin that encapsulates all LTI-related functionality from the View
+    """
+
+    ROLE_MAP = {
+        'student': u'Student',
+        'staff': u'Administrator',
+        'instructor': u'Instructor',
+    }
+    DEFAULT_ROLE = u'Student'
+
+    def get_additional_lti_parameters(self, course: CourseDescriptor, request: HttpRequest) -> Dict[str, str]:
+        lti_config = self.get_lti_config(course)
+        additional_config = lti_config.lti_config.get('additional_parameters', {})
+        return additional_config
+
+    @staticmethod
+    def get_user_id(user: User, course_key: CourseKey):
```

Comment:
It looks like some of these `get_*` helper methods (like this one) _do_ perform quoting on the return value, while others do not.

I'm assuming that's based on the format that `lti_embed` expects of the data?

It's not obvious which do or not, so perhaps it'd make things more clear by either:
- renaming the methods to mention whether quoting is performed, eg: `get_quoted_user_id(...)`
- moving the `quote` call to the invocation of `lti_embed`, like:

```py
@staticmethod
def get_user_id(user: User, course_key: CourseKey):
    return anonymous_id_for_user(user, course_key)

def get_lti_embed_code(...):
    user_id = get_user_id(user, course_key)
    return lti_embed(
        user_id=quote(user_id),
    )
```

My preference would be for the 2nd option.

Reply:
The way I'm looking at this is that this base class is kind of providing a contract for how the subclasses should behave. i.e the fact that some things need to be quoted is based on the value of the thing. So for instance `_get_result_sourcedid` isn't quoted in this case because in the base version we know that it's not required. However, in the case of a subclass it might needs to be quoted because the source data used to build the individual parameters needs to be. 

So I agree the second option is better.